File: android/src/main/kotlin/com/zello/flutter/ZelloPlugin.kt
package com.zello.flutter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Flutter plugin entry point.
 *
 * Bridges Dart -> Zello native SDK via [MethodChannel], and SDK callbacks
 * -> Dart via a single broadcast [EventChannel]. All sink/result callbacks
 * are dispatched onto the platform main thread.
 *
 * The actual Zello SDK calls are isolated behind [ZelloSdkAdapter] so symbol
 * names (which depend on the SDK version you ship) can be corrected in one
 * place without touching channel plumbing.
 */
class ZelloPlugin : FlutterPlugin, MethodCallHandler,
    EventChannel.StreamHandler, ActivityAware {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null

    private var applicationContext: Context? = null
    private var activityBinding: ActivityPluginBinding? = null

    private var adapter: ZelloSdkAdapter? = null
    private var foregroundServiceEnabled = true

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        teardown()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        applicationContext = null
    }

    // -- ActivityAware ---------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityBinding = binding
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding = null
    }

    override fun onDetachedFromActivity() {
        activityBinding = null
    }

    // -- EventChannel.StreamHandler -------------------------------------------

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        adapter?.attachListener(::emit)
    }

    override fun onCancel(arguments: Any?) {
        adapter?.detachListener()
        eventSink = null
    }

    // -- MethodCallHandler -----------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "connect" -> handleConnect(call, result)
            "disconnect" -> safe(result) { adapter?.disconnect(); null }
            "startTalking" -> safe(result) {
                val channel = call.argument<String>("channel")
                    ?: throw IllegalArgumentException("channel is required")
                adapter?.startTalking(channel); null
            }
            "stopTalking" -> safe(result) { adapter?.stopTalking(); null }
            "sendTextMessage" -> safe(result) {
                val channel = call.argument<String>("channel")
                    ?: throw IllegalArgumentException("channel is required")
                val text = call.argument<String>("text")
                    ?: throw IllegalArgumentException("text is required")
                adapter?.sendTextMessage(channel, text); null
            }
            "setStatus" -> safe(result) {
                val status = call.argument<String>("status")
                    ?: throw IllegalArgumentException("status is required")
                adapter?.setStatus(status); null
            }
            "getChannelState" -> safe(result) {
                val channel = call.argument<String>("channel")
                    ?: throw IllegalArgumentException("channel is required")
                adapter?.getChannelState(channel) ?: emptyMap<String, Any?>()
            }
            "dispose" -> safe(result) { teardown(); null }
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(call: MethodCall, result: Result) {
        safe(result) {
            val ctx = applicationContext
                ?: throw IllegalStateException("Plugin not attached to engine")
            val appKey = call.argument<String>("appKey")
                ?: throw IllegalArgumentException("appKey is required")
            foregroundServiceEnabled =
                call.argument<Boolean>("enableForegroundService") ?: true
            val displayName =
                call.argument<String>("channelDisplayName") ?: "Zello"

            adapter?.dispose()
            adapter = ZelloSdkAdapter(
                context = ctx,
                appKey = appKey,
                serviceLabel = displayName,
                useForegroundService = foregroundServiceEnabled,
            ).also { it.attachListener(::emit) }
            null
        }
    }

    private fun handleConnect(call: MethodCall, result: Result) {
        safe(result) {
            val a = adapter ?: throw IllegalStateException("not_initialized")
            val network = call.argument<String>("network")
                ?: throw IllegalArgumentException("network is required")
            val token = call.argument<String>("token")
                ?: throw IllegalArgumentException("token is required")
            val username = call.argument<String>("username")

            if (foregroundServiceEnabled) {
                val ctx = applicationContext!!
                ctx.startForegroundService(
                    Intent(ctx, ZelloForegroundService::class.java)
                )
            }
            a.connect(network, token, username)
            null
        }
    }

    // -- Helpers ---------------------------------------------------------------

    /** Run a block on the platform main thread, mapping exceptions to Result.error. */
    private inline fun safe(result: Result, crossinline block: () -> Any?) {
        mainHandler.post {
            try {
                result.success(block())
            } catch (iae: IllegalArgumentException) {
                result.error("invalid_argument", iae.message, null)
            } catch (ise: IllegalStateException) {
                result.error(ise.message ?: "illegal_state", ise.message, null)
            } catch (t: Throwable) {
                result.error("native_error", t.message ?: t.javaClass.simpleName, null)
            }
        }
    }

    /** Emit a typed event payload to Dart on the main thread. */
    private fun emit(payload: Map<String, Any?>) {
        mainHandler.post { eventSink?.success(payload) }
    }

    private fun teardown() {
        try {
            adapter?.dispose()
        } finally {
            adapter = null
            val ctx = applicationContext
            if (ctx != null && foregroundServiceEnabled) {
                ctx.stopService(Intent(ctx, ZelloForegroundService::class.java))
            }
        }
    }

    companion object {
        // Keep in sync with lib/src/channels.dart
        private const val METHOD_CHANNEL = "com.zello.flutter/method"
        private const val EVENT_CHANNEL = "com.zello.flutter/events"
    }
}

File: ios/zello/Sources/zello/ZelloPlugin.swift
import Flutter
import UIKit
import AVFoundation
#if canImport(PushToTalk)
import PushToTalk
#endif

/// Flutter plugin entry point for iOS.
///
/// Owns the MethodChannel + EventChannel pair, the AVAudioSession
/// configuration, and (when entitled) the Apple PushToTalk `PTChannelManager`.
/// All Zello SDK calls are isolated in `ZelloSdkAdapter` so any SDK-version
/// symbol drift only needs a fix in one file.
public class ZelloPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    // Keep in sync with lib/src/channels.dart
    private static let methodChannelName = "com.zello.flutter/method"
    private static let eventChannelName  = "com.zello.flutter/events"

    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel

    private var eventSink: FlutterEventSink?
    private var adapter: ZelloSdkAdapter?
    private var useApplePTT: Bool = true

    public static func register(with registrar: FlutterPluginRegistrar) {
        let method = FlutterMethodChannel(name: methodChannelName,
                                          binaryMessenger: registrar.messenger())
        let events = FlutterEventChannel(name: eventChannelName,
                                         binaryMessenger: registrar.messenger())
        let plugin = ZelloPlugin(method: method, events: events)
        registrar.addMethodCallDelegate(plugin, channel: method)
        events.setStreamHandler(plugin)
    }

    private init(method: FlutterMethodChannel, events: FlutterEventChannel) {
        self.methodChannel = method
        self.eventChannel = events
        super.init()
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?,
                         eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        adapter?.attachListener { [weak self] payload in
            self?.emit(payload)
        }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        adapter?.detachListener()
        self.eventSink = nil
        return nil
    }

    // MARK: - FlutterPlugin

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "connect":
            handleConnect(call, result: result)
        case "disconnect":
            safe(result) { try self.adapter?.disconnect(); return nil }
        case "startTalking":
            safe(result) {
                let channel: String = try self.requireArg(call, "channel")
                try self.adapter?.startTalking(channel: channel)
                return nil
            }
        case "stopTalking":
            safe(result) { try self.adapter?.stopTalking(); return nil }
        case "sendTextMessage":
            safe(result) {
                let channel: String = try self.requireArg(call, "channel")
                let text: String = try self.requireArg(call, "text")
                try self.adapter?.sendTextMessage(channel: channel, text: text)
                return nil
            }
        case "setStatus":
            safe(result) {
                let status: String = try self.requireArg(call, "status")
                try self.adapter?.setStatus(status)
                return nil
            }
        case "getChannelState":
            safe(result) {
                let channel: String = try self.requireArg(call, "channel")
                return self.adapter?.getChannelState(channel: channel) ?? [:]
            }
        case "dispose":
            safe(result) { self.teardown(); return nil }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method handlers

    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        safe(result) {
            let args = (call.arguments as? [String: Any?]) ?? [:]
            guard let appKey = args["appKey"] as? String else {
                throw ZelloPluginError.invalidArgument("appKey is required")
            }
            self.useApplePTT = (args["enableApplePushToTalk"] as? Bool) ?? true
            let label = (args["channelDisplayName"] as? String) ?? "Zello"

            try self.configureAudioSession()

            self.adapter?.dispose()
            let a = ZelloSdkAdapter(
                appKey: appKey,
                channelDisplayName: label,
                useApplePushToTalk: self.useApplePTT
            )
            a.attachListener { [weak self] payload in self?.emit(payload) }
            self.adapter = a
            return nil
        }
    }

    private func handleConnect(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        safe(result) {
            guard let a = self.adapter else {
                throw ZelloPluginError.illegalState("not_initialized",
                                                    "Zello.initialize() not called")
            }
            let network: String = try self.requireArg(call, "network")
            let token: String = try self.requireArg(call, "token")
            let args = (call.arguments as? [String: Any?]) ?? [:]
            let username = args["username"] as? String
            try a.connect(network: network, token: token, username: username)
            return nil
        }
    }

    // MARK: - Helpers

    private func configureAudioSession() throws {
        let s = AVAudioSession.sharedInstance()
        try s.setCategory(.playAndRecord,
                          mode: .voiceChat,
                          options: [.allowBluetooth,
                                    .allowBluetoothA2DP,
                                    .defaultToSpeaker])
        try s.setActive(true)
    }

    private func requireArg<T>(_ call: FlutterMethodCall, _ key: String) throws -> T {
        let args = (call.arguments as? [String: Any?]) ?? [:]
        guard let v = args[key] as? T else {
            throw ZelloPluginError.invalidArgument("\(key) is required")
        }
        return v
    }

    private func safe(_ result: @escaping FlutterResult,
                      _ block: @escaping () throws -> Any?) {
        DispatchQueue.main.async {
            do {
                result(try block())
            } catch let ZelloPluginError.invalidArgument(msg) {
                result(FlutterError(code: "invalid_argument",
                                    message: msg, details: nil))
            } catch let ZelloPluginError.illegalState(code, msg) {
                result(FlutterError(code: code, message: msg, details: nil))
            } catch {
                result(FlutterError(code: "native_error",
                                    message: error.localizedDescription,
                                    details: nil))
            }
        }
    }

    private func emit(_ payload: [String: Any?]) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(payload)
        }
    }

    private func teardown() {
        adapter?.dispose()
        adapter = nil
        try? AVAudioSession.sharedInstance().setActive(false,
                                                       options: [.notifyOthersOnDeactivation])
    }
}

enum ZelloPluginError: Error {
    case invalidArgument(String)
    case illegalState(String, String)
}
